# Exercises - Week 3 - Dataset - Blackjack

## References
- https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html
- https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html
- https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html

## Contents
1. Data Lab notebooks
2. Two essential transfomers 
1. Dataset intro

## 1. Data Lab notebooks
1. [Python/Libraries/sklearn/Introduction](https://bentley.cloud.databricks.com/#notebook/210807) 
2. [Python/Libraries/sklearn/Preprocessing](https://bentley.cloud.databricks.com/#notebook/404771)

## 2. Two essential transfomers

This section introduces to essential transformers that create numeric features from categorical features and from text features.

There are two types of estimators:
- A _regressor_ fits a target which is continuous and so numeric. (for instance, linear regression)
- A _classifier_ fits a binary or categorical target. (for instance,  logistic regression)

The type of the target variable often determines the type of the estimator (regressor or classifier). 

Most estimators require numeric features.
The remainder of this section introduces transformers which 
- create numeric features from categorical features 
- create numeric features from text features

A single __categorical feature__ with, for instance, three (3) categories/labels would create three (3) binary features. 

The `OneHotEncoder` class from Scikit-learn is designed to transform categorical features into binary features. See:
- https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html

See also the article  [What is One Hot Encoding? Why And When do you have to use it?](https://hackernoon.com/what-is-one-hot-encoding-why-an
d-when-do-you-have-to-use-it-e3c6186d008f) by Rakshith Vasudev of [HackerNoon](https://hackernoon.com/) at [Medium](https://medium.com/).

The image below is from the article linked above.

![](https://cdn-images-1.medium.com/max/2000/1*Ac4z1rWWuU0TzxJRUM62WA.jpeg)

The __vocabulary__ of a __text feature__  is, in our situation, the set of words occuring in (any value of) that feature.
(A set has no duplicates.)

A single text feature with a vocabulary of, for instance, 10,000 words would create 10,000 numeric features.
Each feature corresponds, usually, to a single word. 

There are three basic options for the values of these 10,000 new features:
1. __binary__: the value `1` indicates that the word (corresponding to the column) occurs in the document (corresponding to the row)
1. __frequency__: each value indicates the number of occurences of the word (corresponding to the column) in the document (corresponding to the row)
1. __tf-idf__: each value is the number of occurences of the word (corresponding to the column) in the document (corresponding to the row) divided by the number of documents containing the word (corresponding to the column)

The `CountVectorizer` and `TfidfVectorizer` classes implement these three options. See:
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html
- https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html

See the documentation for an example of each class.

__Exercise:__ Copy, paste and run the `CountVectorizer` example.

In [12]:
from sklearn.feature_extraction.text import CountVectorizer
>>> corpus = [
...     'This is the first document.',
...     'This document is the second document.',
...     'And this is the third one.',
...     'Is this the first document?',
... ]
>>> cvector = CountVectorizer()
>>> X = cvector.fit_transform(corpus)
>>> print(cvector.get_feature_names())
['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']
>>> print(X.toarray())  

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']
[[0 1 1 1 0 0 1 0 1]
 [0 2 0 1 0 1 1 0 1]
 [1 0 0 1 1 0 1 1 1]
 [0 1 1 1 0 0 1 0 1]]

__Exercise:__ Copy, paste and run the `TfidfVectorizer` example.

In [14]:
>>> from sklearn.feature_extraction.text import TfidfVectorizer
>>> corpus = [
...     'This is the first document.',
...     'This document is the second document.',
...     'And this is the third one.',
...     'Is this the first document?',
... ]
>>> vectorizer = TfidfVectorizer()
>>> X = vectorizer.fit_transform(corpus)
>>> print(vectorizer.get_feature_names())
['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']
>>> print(X.shape)

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']
(4, 9)

## 3. Dataset intro

The three datafiles to be used for the project are in an S3 bucket mounted on the directory `/dbfs/mnt/group-ma707`.

In [17]:
%sh ls -hot /dbfs/mnt/group-ma707/data/*

-rw-r--r-- 1 root 259K Jan 29 17:44 /dbfs/mnt/group-ma707/data/5tc_plus_ind_vars.csv
-rw-r--r-- 1 root 12M Jan 29 17:44 /dbfs/mnt/group-ma707/data/mining_com_coal.csv
-rw-r--r-- 1 root 11M Jan 29 17:44 /dbfs/mnt/group-ma707/data/mining_com_iron_ore.csv

The report will investigate at least two datasets. For each we need to do the following:
1. find or create a target variable (there are at least two)
1. find or create features to use in predicting the target variable(s)

To start, display the first few lines of the datafiles to see what we have to work with.

The first three lines of the `5tc_plus_ind_vars.csv` file:

In [20]:
%sh head -n 3 /dbfs/mnt/group-ma707/data/5tc_plus_ind_vars.csv

"Date","BCI","C5","C7","P1A_03","P2A_03","P4_03","P3A~IV","SHFE_AL3","RICI","ICE_KC3","CME_SM3","CME_LC2","OPEC_ORB","SHFE_CU3","CME_LN1","CME_FC3","P3A_03","SHFE_RB3","CME_S2","ICE_SB3","CME_LN3","CME_LN2","ICE_TIB3","ICE_TIB4","BCI_5TC"
2011-12-05,3390,12.529,15.718,15329,24935,3324,16.54,16.54,3688.22,3688.22,3688.22,3688.22,110.3,110.3,110.3,110.3,10503,10503,1136.5,1136.5,1136.5,1136.5,-6.66,-6.6,29966
2011-12-06,3387,12.379,15.705,15328,24967,3288,16.55,16.55,3700.6,3700.6,3700.6,3700.6,109.49,109.49,109.49,109.49,10412,10412,1139.5,1139.5,1139.5,1139.5,-6.84,-6.77,29990

The last column (`BCI_5TC`) is one of our target variables. 

Notice the `Date` field. The dataframe produced with the above target variable will be a time series (each record has data from a specific date). The target variable will be predicted with data from previous dates/times. We will write transformer classes to create features that are lagged values of these columns.

The target variable will be predicted from these lagged values. Features from the other files can also be used.

__Exercise:__ 
- Read in this file using the pandas `read_csv` function. 
- Store the resulting dataframe in `bci_5tc_plus_pdf`.

Here we import the pandas library into python and read in the specified csv file.  There are no other steps needed here because the csv file does not contain data that requires special coding

In [24]:
import pandas as pd
bci_5tc_plus_pdf = pd.read_csv('/dbfs/mnt/group-ma707/data/5tc_plus_ind_vars.csv')

The first nine lines of the `mining_com_coal.csv` file:

In [26]:
%sh head -n 9 /dbfs/mnt/group-ma707/data/mining_com_coal.csv

content,date,origin,subtitle,tags,title,url
"Electric cars are better for the environment than traditional gasoline models, and that benefit will grow as power generators shift away from coal.
That’s the conclusion of research by BloombergNEF, which found carbon dioxide emissions from battery-powered vehicles were about 40 percent lower than for internal combustion engines last year. The difference was biggest in Britain and the U.K., which have large renewables industries. It still held in China, which is more reliant on coal to make electricity. Carbon dioxide emissions from battery-powered vehicles were about 40 percent lower than for internal combustion engines last yearThe report adds clarity to the debate about the lifetime emissions of electric vehicles, which while they don’t pollute on the road do consume electricity that is often generated using fossil fuels. BNEF’s research assumes that electric cars will become cleaner in the coming years as utilities close coal plants and draw more energy from wind and solar farms, a process well underway almost everywhere except Southeast Asia.
“When an internal combustion vehicle rolls off the line its emissions per km are set, but for an EV they keep falling every year as the grid gets cleaner,” Colin McKerracher, a transport analyst at BNEF said.
The chart above shows BNEF’s forecasts for how much the gap will grow, with China expected to make rapid progress as its renewables industry takes a bigger share of the power generation market in the coming years.
The widespread adoption of renewables will decrease average emissions by as much as 90 percent in the U.K. and over a third in Japan out to 2040, according to BNEF.
The global share of zero-carbon electricity generation is set to increase from 38 percent last year to 63 percent by 2040, according to projections from BNEF. While technological improvements will see related emissions from combustion engines falling by about 1.9 percent a year through to 2040, pollution from electric vehicles will fall anywhere from 3 percent and 10 percent annually. That’s largely because of grid decarbonization but also reduced electricity consumption, BNEF says.
(By Jeremy Hodges)",2019-01-15T23:04:53+00:00,bloomberg-news,,britain china coal cobalt copper europe graphite lithium nickel u-s-a,Electric cars seen getting cleaner even where grids rely on coal,http://www.mining.com/web/electric-cars-seen-getting-cleaner-even-grids-rely-coal/
"Project Halo, the South African consortium selected as the preferred bidder in the highly contested acquisition of coal mines previously owned by the Gupta family, is denying circulating allegations of favouritism for allegedly being linked to the same family.

All columns from this file can be used to produce features for the `BCI_5TC` target variable.

__Exercise:__ 
- Read in this file using the pandas `read_csv` function.
- You will encounter an error. Search to find the single parameter to avoid the error. 
- Store the resulting dataframe in `mining_com_coal_pdf`.

The next two lines of code take the csv files and encode them usng the "ISO..." method relatiive to the dataset.  This method allows all the data to be read into the dataframe correctly,  but will require some modifications later on for use in the subsequent exercises.

In [30]:
import pandas as pd
mining_com_coal_pdf=pd.read_csv('/dbfs/mnt/group-ma707/data/mining_com_coal.csv', encoding = "ISO-8859-1")
mining_com_coal_pdf.head()

Out[ 17 ]: 
 content ... url
0 Electric cars are better for the environment t... ... http://www.mining.com/web/electric-cars-seen-g...
1 Project Halo, the South African consortium sel... ... http://www.mining.com/project-halo-rejects-fav...
2 * China's 2018 iron ore imports fall to 1.064 ... ... http://www.mining.com/web/chinas-2018-iron-ore...
3 Angela Merkel this week will seek to bridge di... ... http://www.mining.com/web/merkel-seeks-heal-ri...
4 A coal mine roof collapse in northwest China's... ... http://www.mining.com/web/coal-mine-collapse-c...

[5 rows x 7 columns]

__Exercise:__ 
- Read in the other file (not displayed above, but in the same directory) using the pandas `read_csv` function.
- You will encounter an error. Try the fix you found above. 
- Store the resulting dataframe in `mining_com_iron_ore_pdf`.

See comments above - the same problem is found and the same solution is used.

In [33]:
import pandas as pd
mining_com_iron_ore_pdf = pd.read_csv ('/dbfs/mnt/group-ma707/data/mining_com_iron_ore.csv', encoding = "ISO-8859-1")
mining_com_iron_ore_pdf[['origin','subtitle']]

Out[ 5 ]: 
 origin subtitle
0 bloomberg-news NaN
1 reuters NaN
2 reuters NaN
3 reuters NaN
4 mining.com-editor NaN
5 bloomberg-opinion NaN
6 reuters NaN
7 reuters NaN
8 cecilia-jamasmie NaN
9 reuters NaN
10 bloomberg-news NaN
11 reuters NaN
12 reuters NaN
13 cecilia-jamasmie NaN
14 bloomberg-news NaN
15 bloomberg-news NaN
16 reuters NaN
17 reuters NaN
18 reuters NaN
19 frik-els NaN
20 reuters NaN
21 reuters NaN
22 frik-els NaN
23 reuters NaN
24 cecilia-jamasmie NaN
25 northern-miner-staff NaN
26 northern-miner-staff NaN
27 bloomberg-news NaN
28 reuters NaN
29 cecilia-jamasmie NaN
... ... ...
3997 mining.com-editor NaN
3998 engineering-news NaN
3999 infomine---mining-equipment-and-supplier-news NaN
4000 infomine---mining-equipment-and-supplier-news NaN
4001 infomine---mining-equipment-and-supplier-news NaN
4002 infomine---mining-equipment-and-supplier-news NaN
4003 infomine---mining-equipment-and-supplier-news NaN
4004 mining-news NaN
4005 dan-oancea NaN
4006 mining-news NaN
4007 mining-news NaN
4008 infomine---mining-equipment-and-supplier-news NaN
4009 mining-news NaN
4010 infomine---mining-equipment-and-supplier-news NaN
4011 mining-news NaN
4012 international-mining NaN
4013 infomine---mining-equipment-and-supplier-news NaN
4014 mining-news NaN
4015 mining.com-editor NaN
4016 infomine---mining-equipment-and-supplier-news NaN
4017 infomine---mining-equipment-and-supplier-news NaN
4018 infomine---mining-equipment-and-supplier-news NaN
4019 infomine---mining-equipment-and-supplier-news NaN
4020 infomine---mining-equipment-and-supplier-news NaN
4021 infomine---mining-equipment-and-supplier-news NaN
4022 infomine---mining-equipment-and-supplier-news NaN
4023 infomine---mining-equipment-and-supplier-news NaN
4024 infomine---mining-equipment-and-supplier-news NaN
4025 infomine---mining-equipment-and-supplier-news NaN
4026 infomine---mining-equipment-and-supplier-news NaN

[4027 rows x 2 columns]

__Exercise:__ 
- Reset the column names of each dataframe to use lower case and snake case. 
- Display the column attribute of each dataframe.

The following commands change the column names to be lower case (denoted in the coding by 'lower') ad to use snake case by replacing all spaces between words with underscores. This allows the code output to be easily read and interpreted by another programmer.

In [36]:
mining_com_iron_ore_pdf.columns =  mining_com_iron_ore_pdf.columns.str.lower().str.replace(' ','_')
mining_com_coal_pdf.columns = mining_com_coal_pdf.columns.str.lower().str.replace(' ','_')
bci_5tc_plus_pdf.columns = bci_5tc_plus_pdf.columns.str.lower().str.replace(' ','_')
bci_5tc_plus_pdf.info()
bci_5tc_plus_pdf.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1602 entries, 0 to 1601
Data columns (total 26 columns):
date 1602 non-null object
bci 1602 non-null int64
c5 1602 non-null float64
c7 1602 non-null float64
p1a_03 1602 non-null int64
p2a_03 1602 non-null int64
p4_03 1602 non-null int64
p3a~iv 1602 non-null float64
shfe_al3 1602 non-null float64
rici 1602 non-null float64
ice_kc3 1602 non-null float64
cme_sm3 1602 non-null float64
cme_lc2 1602 non-null float64
opec_orb 1602 non-null float64
shfe_cu3 1602 non-null float64
cme_ln1 1602 non-null float64
cme_fc3 1602 non-null float64
p3a_03 1602 non-null int64
shfe_rb3 1602 non-null int64
cme_s2 1602 non-null float64
ice_sb3 1602 non-null float64
cme_ln3 1602 non-null float64
cme_ln2 1602 non-null float64
ice_tib3 1602 non-null float64
ice_tib4 1602 non-null float64
bci_5tc 1602 non-null int64
dtypes: float64(18), int64(7), object(1)
memory usage: 325.5+ KB
 Out[ 39 ]: 
 date bci c5 c7 ... cme_ln2 ice_tib3 ice_tib4 bci_5tc
0 2011-12-05 3390 12.529 15.718 ... 1136.50 -6.66 -6.60 29966
1 2011-12-06 3387 12.379 15.705 ... 1139.50 -6.84 -6.77 29990
2 2011-12-07 3405 12.163 15.923 ... 1141.00 -6.37 -6.27 30337
3 2011-12-08 3529 12.421 16.286 ... 1142.25 -7.10 -7.00 31803
4 2011-12-09 3697 13.325 16.486 ... 1116.50 -7.43 -7.37 33276

[5 rows x 26 columns]

__Exercise:__ Check each dataframe to ensure that it looks reasonable, using the `info`, `head` and `tail` methods.

We did not soend much time here, as we reached the conclusion that the dataframes looked reasonable after a couple of iterations through the workbook.  We noted the usefulness of the info method particularly, and discussed how this might be used later on when determining our prooject dataset.

In [39]:
mining_com_iron_ore_pdf.head()
mining_com_iron_ore_pdf.tail()
mining_com_coal_pdf.head()
bci_5tc_plus_pdf.head()

Out[ 20 ]: 
 date bci c5 c7 ... cme_ln2 ice_tib3 ice_tib4 bci_5tc
0 2011-12-05 3390 12.529 15.718 ... 1136.50 -6.66 -6.60 29966
1 2011-12-06 3387 12.379 15.705 ... 1139.50 -6.84 -6.77 29990
2 2011-12-07 3405 12.163 15.923 ... 1141.00 -6.37 -6.27 30337
3 2011-12-08 3529 12.421 16.286 ... 1142.25 -7.10 -7.00 31803
4 2011-12-09 3697 13.325 16.486 ... 1116.50 -7.43 -7.37 33276

[5 rows x 26 columns]

__Exercise:__ Notice that each file/dataframe has a date variable. 
- Use these variables to join these dataframes (created from these files) into a single dataframe. 
- Use a left join with `bci_5tc_plus_pdf` as your base dataframe. 
- Store your result in `bci_5tc_plus_mining_com_pdf`.

The following code takes the dataframes specified above, and performs a left merge on them.  Because the iron ore dataframe and the coal dataframe dates were substantially longer (due to timestamps) than the bci dataframe, the dates were sliced to represnt values from 0-10, elimiinating the time portion of the date and allowing all dates to match correctly.

In [42]:
mining_com_iron_ore_pdf['date']=mining_com_iron_ore_pdf['date'].str.slice(0,10)
mining_com_coal_pdf['date']=mining_com_coal_pdf['date'].str.slice(0,10)

In [43]:
bci_5tc_plus_mining_com_pdf1=pd.merge(bci_5tc_plus_pdf,mining_com_iron_ore_pdf,how='left',on=['date', 'date'])
bci_5tc_plus_mining_com_pdf=pd.merge(bci_5tc_plus_mining_com_pdf1,mining_com_coal_pdf,how='left',on=['date', 'date'])

__Exercise:__ Check the result of your join to see if there were any rows in the base file that did not match with a row in either of the two _mining_ dataframes.

bci_5tc_plus_pdf has 1602 entries and bci_5tc_plus_mining_com_pdf has 6665 entries. It means that some rows in the base file matches mutiple rows in other two mining dataframes. There are some missing values in content_x and content_y, which means some rows in the case didn't match with a row in either of the two mining dataframes

In [46]:
bci_5tc_plus_mining_com_pdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6665 entries, 0 to 6664
Data columns (total 38 columns):
date 6665 non-null object
bci 6665 non-null int64
c5 6665 non-null float64
c7 6665 non-null float64
p1a_03 6665 non-null int64
p2a_03 6665 non-null int64
p4_03 6665 non-null int64
p3a~iv 6665 non-null float64
shfe_al3 6665 non-null float64
rici 6665 non-null float64
ice_kc3 6665 non-null float64
cme_sm3 6665 non-null float64
cme_lc2 6665 non-null float64
opec_orb 6665 non-null float64
shfe_cu3 6665 non-null float64
cme_ln1 6665 non-null float64
cme_fc3 6665 non-null float64
p3a_03 6665 non-null int64
shfe_rb3 6665 non-null int64
cme_s2 6665 non-null float64
ice_sb3 6665 non-null float64
cme_ln3 6665 non-null float64
cme_ln2 6665 non-null float64
ice_tib3 6665 non-null float64
ice_tib4 6665 non-null float64
bci_5tc 6665 non-null int64
content_x 6210 non-null object
origin_x 6225 non-null object
subtitle_x 0 non-null float64
tags_x 6225 non-null object
title_x 6225 non-null object
url_x 6225 non-null object
content_y 6271 non-null object
origin_y 6275 non-null object
subtitle_y 0 non-null float64
tags_y 6275 non-null object
title_y 6275 non-null object
url_y 6276 non-null object
dtypes: float64(20), int64(7), object(11)
memory usage: 2.0+ MB

From the `bci_5tc_plus_mining_com_pdf` dataframe, will create a dataframe with features to predict the `bci_5tc`. 

The remaining exercises create another dataframe, that will contain a different target.

__Exercise:__ 
- Use the `CountVectorizer` class to create numeric columns from each of the three text columns in the `mining_com_iron_ore_pdf` dataframe. 
- Display the shape of each of results of the `transform` method for each of these text columns.

The following code utilizes the countectorizer methods descrbed above, note the need top transform the values of the content, origin, and tags columns away from ISO to Unicode (specified by U). After this, the shape of all the columns are shown, we note that the shapes all differ, because the tags and origin values have less unique information than the overall content

In [50]:
mining_content = cvector.fit_transform(mining_com_iron_ore_pdf['content'].values.astype('U'))
mining_origin = cvector.fit_transform(mining_com_iron_ore_pdf['origin'].values.astype('U'))
mining_tags = cvector.fit_transform (mining_com_iron_ore_pdf['tags'].values.astype('U'))

In [51]:
mining_content.shape,mining_origin.shape,mining_tags.shape

Out[ 58 ]: ((4027, 39727), (4027, 455), (4027, 728))

__Exercise:__ produce a count of the unique values of each of the `origin` columns from each of the _mining_ dataframes.
(produce two results)

The following code looks to utilize the count vectorize method defined above, and then utilizes the get feture names associated with the function.  By using. the length of the get features name we can see the number of unique values in the two dataframes.

In [54]:
print(len(set(mining_com_coal_pdf['origin'].values)))
print(len(set(mining_com_iron_ore_pdf['origin'].values)))

349
257

__Exercise:__ from one of the _mining_ dataframes:
- transform the `origin` column using the `OneHotEncoder` class 
- display the shape of the result

In [56]:
>>> from sklearn.preprocessing import OneHotEncoder,LabelEncoder
>>> enc = OneHotEncoder(handle_unknown='ignore')

In order to use the one hot encoder method we needed to determine numeric values for the different origins.  In order to do thiis we decided to utiliize the pandas function 'get dummies' which created dumy variables for each origin.  We were then able to use our encoder method, and generate a sparse array as defined, and count the different values using the shape function.

In [58]:
le = LabelEncoder()
enc= OneHotEncoder()
tmp=le.fit_transform(mining_com_coal_pdf['origin'].values.tolist())
x_train = enc.fit_transform(tmp.reshape(-1, 1))
x_train.shape

/databricks/python/lib/python3.5/site-packages/sklearn/preprocessing/_encoders.py:371: FutureWarning: The handling of integer data will change in version 0.22. Currently, the categories are determined based on the range [0, max(values)], while in the future they will be determined based on the unique values.
If you want the future behaviour and silence this warning, you can specify "categories='auto'".
In case you used a LabelEncoder before this OneHotEncoder to convert the categories to integers, then you can now use the OneHotEncoder directly.
 warnings.warn(msg, FutureWarning)
 Out[ 71 ]: (4460, 349)

We note that within the one hot encoder method shown above we get a result of 349 unique values, indicating that the one hot encoder method provided a similar result to that of the unique count methods we performed earlier.

__The End__